Added gradle tasks for counting methods

ricellis opened the pull request:
*What*
Added gradle tasks for counting methods

*Why*
Older versions of Android have a 65k method limit and it is good to know how much the library might take up.

*How*
Added tasks to copy all dependencies and produce a single dex file.
Extracted the `method id` statistic from the dex log.
Added task to delete the copied dependencies and generated file.

*Test*
No new tests, the task is designed to be run manually when a new method count is needed and I have tested it locally.
Runs with e.g.:
`./gradlew methodCount -Ddx.exec=~/Library/Android/sdk/build-tools/24.0.0/dx`
with the `dx.exec` property pointing to the Android build tools dx executable.

Fixes #271

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,36 @@ subprojects {
+        """.replace("dx.exec",(System.getProperty("dx.exec") == null) ? "dx" : System.getProperty("dx.exec"))
+    }
+
+    task deleteDex(type: Delete, dependsOn: dx) {
```
I think you drop the `deleteDex` class and rely on the `clean` targets doing the right thing.

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,36 @@ subprojects {
+
+    task copyDependencies(type: Copy) {
+        from configurations.runtime
+        into new File(buildDir, "dex")
```
`File(buildDir, "dex-all")` should be in some shared config and re-used by these tasks. I suggest something like `dex-all` just to distinguish from other invocations of `dex`

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,36 @@ subprojects {
+    }
+
+    task copyLib(type: Copy, dependsOn: copyDependencies) {
+        from new File(buildDir, "libs")
```
I wonder if there is a config variable for the `"libs"` dir, but we already use this string elsewhere so it's already established practice.

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,36 @@ subprojects {
+        into new File(buildDir, "dex")
+    }
+
+    task dx(type: Exec, dependsOn: copyLib) {
```
maybe call this task `dexAll`?

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,36 @@ subprojects {
+
+    task methodCount(dependsOn: [build, copyDependencies, copyLib, dx, deleteDex]) {
+        // Run this task to get the Android method counts.
+        // Relies on the dx.exec system property being set to point to the dx executable
```
or on `dx` being in your `PATH`

ricellis commented:
@tomblench addressed comments in 0bf7f29:

Added constant `dexAllDir` for `dex-all` directory.
Used `jar.destinationDir` instead of buildDir + “libs”.
Renamed `dx` task to `dexAll`.
Added clarification that `dx` can be on the path for gradle.
Removed `deleteDex` task, `clean` will remove it anyway.

tomblench commented on build.gradle:
```diff
@@ -224,6 +224,33 @@ subprojects {
+        """.replace("dx.exec",(System.getProperty("dx.exec") == null) ? "dx" : System.getProperty("dx.exec"))
+    }
+
+    task methodCount(dependsOn: [build, copyDependencies, copyLib, dexAll]) {
```
`copyDependencies, copyLib` dependencies are implied don't strictly need to be there.

ricellis replied:
@tomblench fixed the dependencies and made these other improvements in 42dbe4b:
* Created `AndroidExec` task type to add android tools to the path reading
`sdk.home` or `ANDROID_HOME`.
* Removed `dx.exec` property and used `AndroidExec` type instead.
* Fixed dependencies including using `assemble` instead of `build` to avoid
running tests if just doing `methodcount`.
* Added input/output directories to `dexAll` task to avoid running twice
unnecessarily.

Unfortunately sharing the `AndroidExec` type with the AndroidTest/app project is proving problematic because they don't share a root build. I think I'd like to merge what I have here and consider the sharing later (possibly with the log collection as that is going to need android commands as well).

tomblench commented:
+1
